src/Neo/SmartContract/ApplicationEngine.cs

```diff
@@ -612,11 +612,30 @@ public void SetState<T>(T state)
 
         public bool IsHardforkEnabled(Hardfork hardfork)
         {
-            if (PersistingBlock is null)
-                return true;
-            if (!ProtocolSettings.Hardforks.TryGetValue(hardfork, out uint height))
+            // Get all the known hardforks
+            var allHardforks = Enum.GetValues(typeof(Hardfork)).Cast<Hardfork>().ToList();
+            allHardforks.Sort(); // Ensure they are sorted in the order of their definition
+
+            // Case 1: If no hardfork is specified in the configuration, enable all hardforks by default
+            if (PersistingBlock is null || ProtocolSettings.Hardforks.Count == 0)
                 return true;
-            return PersistingBlock.Index >= height;
+
+            // If the hardfork is not specified in the configuration, check if it's a new hardfork. If yes, disable it.
+            if (!ProtocolSettings.Hardforks.ContainsKey(hardfork))
+            {
+                // If the index of the hardfork in the list of known hardforks is greater than the index of the last hardfork in the configuration, disable it.
+                if (allHardforks.IndexOf(hardfork) > allHardforks.IndexOf(ProtocolSettings.Hardforks.Keys.Last()))
+                    return false;
+            }
+
+            uint height;
+            if (ProtocolSettings.Hardforks.TryGetValue(hardfork, out height))
+            {
+                // If the hardfork is set to a certain height in the configuration, check if the current block height is greater than or equal to this height
+                return PersistingBlock.Index >= height;
+            }
+            // For those old hardforks that are not explicitly specified in the configuration, we enable them by default at height 0.
+            return true;
```

Comment:
Can we add a check to https://github.com/neo-project/neo/blob/d63eb1da7a99e7459a647dc3355145f5cccc6d74/src/Neo/ProtocolSettings.cs#L185 (or to some other place) to explicitly forbid situations when in JSON config we have something like:
```
HF_A: 100
HF_C: 123
```
This situation is described in https://github.com/neo-project/neo/pull/2886#issuecomment-1674756214, and we should treat such node config as invalid and probably throw an exception in this case.

Reply:
I personaly dont think we should handle those malicious or erronious configurations as it is the responsibility of the user to make it right, but anyway, please check the updated code.

Reply:
Then we probably need some third opinion. @shargon, @roman-khimov, maybe someone else, what do you think on this topic?

Reply:
Resolved, see the updated code.